PalasoUIWindowsForms/Keyboarding/Windows/WinKeyboardAdaptor.cs

```diff
@@ -48,14 +49,124 @@ public LayoutName(string layout, string localizedLayout)
 			public string LocalizedName;
 		}
 
+		/// <summary>
+		/// This class receives notifications from TSF when the input method changes.
+		/// It also implements a fallback to Windows messages if TSF isn't available,
+		/// e.g. on Windows XP.
+		/// </summary>
+		private class TfLanguageProfileNotifySink : ITfLanguageProfileNotifySink
+		{
+			private readonly WinKeyboardAdaptor _keyboardAdaptor;
+			private List<Form> _toplevelForms = new List<Form>();
+
+			public TfLanguageProfileNotifySink(WinKeyboardAdaptor keyboardAdaptor)
+			{
+				_keyboardAdaptor = keyboardAdaptor;
+			}
+
+			#region ITfLanguageProfileNotifySink Members
+
+			public bool OnLanguageChange(ushort langid)
+			{
+				// In my tests we never hit this method (Windows 8.1). I don't know if the
+				// method signature is wrong or why that is.
+				return true;
+			}
+
+			public void OnLanguageChanged()
+			{
+				var winKeyboard = KeyboardForCurrentInputProfile;
+				Debug.WriteLine("Language changed from {0} to {1}",
+					Keyboard.Controller.ActiveKeyboard != null ? Keyboard.Controller.ActiveKeyboard.Layout : "<null>",
+					winKeyboard != null ? winKeyboard.Layout : "<null>");
+
+				_keyboardAdaptor.m_windowsLanguageProfileSinks.ForEach(
+					(sink) => sink.OnInputLanguageChanged(Keyboard.Controller.ActiveKeyboard, winKeyboard));
+			}
+
+			private WinKeyboardDescription KeyboardForCurrentInputProfile
+			{
+				get
+				{
+					if (_keyboardAdaptor.ProfileMgr != null)
+					{
+						var profile = _keyboardAdaptor.ProfileMgr.GetActiveProfile(Guids.TfcatTipKeyboard);
+						return Keyboard.Controller.AllAvailableKeyboards.OfType<WinKeyboardDescription>()
+							.FirstOrDefault(winKeybd => InputProcessorProfilesEqual(profile, winKeybd.InputProcessorProfile));
+					}
+
+					// Probably Windows XP where we don't have ProfileMgr
+					var lang = _keyboardAdaptor.ProcessorProfiles.GetCurrentLanguage();
+					return Keyboard.Controller.AllAvailableKeyboards.OfType<WinKeyboardDescription>()
+						.FirstOrDefault(winKeybd => winKeybd.InputProcessorProfile.LangId == lang);
+				}
+			}
+
+			private static bool InputProcessorProfilesEqual(TfInputProcessorProfile profile1, TfInputProcessorProfile profile2)
+			{
+				// Don't compare Flags - they can be different and it's still the same profile
+				return profile1.ProfileType == profile2.ProfileType &&
+					profile1.LangId == profile2.LangId &&
+					profile1.ClsId == profile2.ClsId &&
+					profile1.GuidProfile == profile2.GuidProfile &&
+					profile1.CatId == profile2.CatId &&
+					profile1.HklSubstitute == profile2.HklSubstitute &&
+					profile1.Caps == profile2.Caps &&
+					profile1.Hkl == profile2.Hkl;
+			}
+			#endregion
+
+			#region Fallback if TSF isn't available
+			private void OnWindowsMessageInputLanguageChanged(object sender,
+				InputLanguageChangedEventArgs inputLanguageChangedEventArgs)
```

Comment:
Never mind, I see you only register this handler if TSF isn't available. Might be worth enhancing the comment to say that if you do a revision.